File: app/routers/budget.py
"""Budget endpoints — franjas, onboarding, history."""
from datetime import date
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session
from ..database import get_db
from ..models import User, Transaction, Alert
from ..security import get_current_user
from ..services.splits import expense_amount, reimbursement_map
from ..services.date_utils import days_in_month as _days_in_month

router = APIRouter(prefix="/api/budget", tags=["budget"])


class OnboardingPayload(BaseModel):
    name: str
    monthly_income: float
    necesidades_pct: int = 50
    gustos_pct: int = 30
    ahorro_pct: int = 20
    payday: int = 1
    income_is_variable: bool = False


class BudgetUpdatePayload(BaseModel):
    monthly_income: Optional[float] = None
    necesidades_pct: Optional[int] = None
    gustos_pct: Optional[int] = None
    ahorro_pct: Optional[int] = None
    payday: Optional[int] = None
    income_is_variable: Optional[bool] = None


class BalancePayload(BaseModel):
    balance: float


def _franja_data(user: User, txs: list, month: str, days_remaining: int = 1,
                 reimb: dict | None = None, income_base: float | None = None) -> dict:
    income = income_base if income_base is not None else (user.monthly_income or 0)
    limits = {
        "necesidades": income * user.necesidades_pct / 100,
        "gustos": income * user.gustos_pct / 100,
        "ahorro": income * user.ahorro_pct / 100,
    }
    reimb = reimb or {}
    visible = [t for t in txs if not getattr(t, 'is_internal_transfer', False) and not getattr(t, 'is_duplicate', False)]
    spent = {
        cat: sum(expense_amount(t, reimb) for t in visible if t.category == cat and getattr(t, 'tx_type', 'expense') == 'expense')
        for cat in limits
    }
    dr = max(days_remaining, 1)
    return {
        "month": month,
        "income": income,
        "franjas": [
            {
                "name": cat,
                "label": {"necesidades": "Necesidades", "gustos": "Gustos", "ahorro": "Ahorro"}[cat],
                "pct_config": getattr(user, f"{cat}_pct"),
                "limit": limits[cat],
                "spent": spent[cat],
                "remaining": max(0, limits[cat] - spent[cat]),
                "usage_pct": round(spent[cat] / limits[cat] * 100, 1) if limits[cat] > 0 else 0,
                "daily_allowance": round(max(0, limits[cat] - spent[cat]) / dr, 0),
            }
            for cat in ["necesidades", "gustos", "ahorro"]
        ],
    }


@router.get("/current")
def get_current_budget(month: Optional[str] = None, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    from ..services.recurring import apply_recurring

    now = date.today()
    current_month = now.strftime("%Y-%m")
    if month:
        try:
            year, mo = (int(p) for p in month.split("-"))
            target = date(year, mo, 1)
        except (ValueError, TypeError):
            raise HTTPException(400, "Mes inválido, usá el formato YYYY-MM")
        month = target.strftime("%Y-%m")
    else:
        month = current_month
        target = date(now.year, now.month, 1)

    is_current = month == current_month
    if is_current:
        apply_recurring(db, user.id)

    txs = db.query(Transaction).filter(
        Transaction.user_id == user.id,
        Transaction.month == month,
        Transaction.status.in_(["confirmed", "classified"])
    ).all()

    days_in_month = _days_in_month(target.year, target.month)
    days_passed = now.day if is_current else days_in_month

    days_remaining = max(days_in_month - days_passed, 1) if is_current else 0
    reimb = reimbursement_map(db, user.id)

    visible_txs = [t for t in txs if not getattr(t, 'is_internal_transfer', False) and not getattr(t, 'is_duplicate', False)]
    tracked_income = sum(t.amount for t in visible_txs
                         if getattr(t, 'tx_type', 'expense') == 'income'
                         and not getattr(t, 'is_reimbursement', False))
    total_expenses = sum(expense_amount(t, reimb) for t in visible_txs
                         if getattr(t, 'tx_type', 'expense') == 'expense')

    is_variable = bool(getattr(user, 'income_is_variable', False))
    declared_income = user.monthly_income or 0

    # Base del presupuesto:
    #  - Ingreso fijo: el sueldo declarado es el piso; si lo registrado lo supera, gana lo real.
    #  - Ingreso variable: no hay sueldo fijo, así que la base son los ingresos
    #    realmente registrados este mes (MP/Gmail/manual).
    if is_variable:
        total_income = tracked_income
        income_base = tracked_income
    else:
        total_income = max(tracked_income, declared_income)
        income_base = total_income

    data = _franja_data(user, txs, month, days_remaining=days_remaining, reimb=reimb, income_base=income_base)
    data["days_passed"] = days_passed
    data["days_in_month"] = days_in_month
    data["days_remaining"] = days_in_month - days_passed

    pending_count = sum(1 for t in txs if t.needs_review and t.status != "reviewed")

    data["month"] = month
    data["is_current_month"] = is_current
    data["total_income"] = total_income
    data["tracked_income"] = tracked_income
    data["declared_income"] = declared_income
    data["income_is_variable"] = is_variable
    data["income_is_declared"] = (not is_variable) and declared_income > tracked_income
    data["total_expenses"] = total_expenses
    data["balance"] = user.balance or 0.0
    data["pending_count"] = pending_count
    data["onboarding_done"] = user.onboarding_done
    data["name"] = user.name
    data["payday"] = user.payday

    alerts = db.query(Alert).filter(
        Alert.user_id == user.id,
        Alert.is_read == False
    ).order_by(Alert.created_at.desc()).all()

    data["alerts"] = [
        {
            "id": a.id,
            "type": a.type,
            "category": a.category,
            "message": a.message,
            "ai_advice": a.ai_advice,
            "severity": a.severity,
            "payload": a.payload,
            "created_at": a.created_at.isoformat(),
        }
        for a in alerts
    ]

    return data


@router.post("/onboarding")
def complete_onboarding(payload: OnboardingPayload, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    if payload.necesidades_pct + payload.gustos_pct + payload.ahorro_pct != 100:
        raise HTTPException(400, "Los porcentajes deben sumar 100")

    user.name = payload.name
    user.monthly_income = payload.monthly_income
    user.necesidades_pct = payload.necesidades_pct
    user.gustos_pct = payload.gustos_pct
    user.ahorro_pct = payload.ahorro_pct
    user.payday = payload.payday
    user.income_is_variable = payload.income_is_variable
    user.onboarding_done = True
    db.commit()
    return {"ok": True}


@router.patch("/settings")
def update_budget_settings(payload: BudgetUpdatePayload, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    if payload.monthly_income is not None:
        user.monthly_income = payload.monthly_income
    if payload.necesidades_pct is not None:
        user.necesidades_pct = payload.necesidades_pct
    if payload.gustos_pct is not None:
        user.gustos_pct = payload.gustos_pct
    if payload.ahorro_pct is not None:
        user.ahorro_pct = payload.ahorro_pct
    if payload.payday is not None:
        user.payday = payload.payday
    if payload.income_is_variable is not None:
        user.income_is_variable = payload.income_is_variable

    total = user.necesidades_pct + user.gustos_pct + user.ahorro_pct
    if total != 100:
        db.rollback()
        raise HTTPException(400, f"Los porcentajes suman {total}, deben ser 100")

    db.commit()
    return {"ok": True}


@router.patch("/balance")
def update_balance(payload: BalancePayload, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    """Ajuste manual del balance de caja (ej: efectivo, correcciones)."""
    user.balance = payload.balance
    db.commit()
    return {"ok": True, "balance": user.balance}


@router.get("/months")
def get_available_months(db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    """Meses con datos reales del usuario, más el mes actual (aunque esté vacío)."""
    rows = db.query(Transaction.month).filter(
        Transaction.user_id == user.id,
        Transaction.status.in_(["confirmed", "classified"])
    ).distinct().all()
    months = {r[0] for r in rows if r[0]}
    months.add(date.today().strftime("%Y-%m"))
    return sorted(months, reverse=True)


@router.post("/alerts/{alert_id}/read")
def mark_alert_read(alert_id: int, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    alert = db.query(Alert).filter_by(id=alert_id, user_id=user.id).first()
    if not alert:
        raise HTTPException(404, "Alerta no encontrada")
    alert.is_read = True
    db.commit()
    return {"ok": True}


@router.get("/history")
def get_budget_history(db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    months_raw = db.query(Transaction.month).filter(
        Transaction.user_id == user.id,
        Transaction.status.in_(["confirmed", "classified"])
    ).distinct().all()

    months = sorted([m[0] for m in months_raw], reverse=True)[:6]
    result = []
    for month in months:
        txs = db.query(Transaction).filter(
            Transaction.user_id == user.id,
            Transaction.month == month,
            Transaction.status.in_(["confirmed", "classified"])
        ).all()
        result.append(_franja_data(user, txs, month, reimb=reimbursement_map(db, user.id)))

    return result

File: services/budget-service/src/routes/budget.js
/** Budget endpoints — franjas, onboarding, history. */
import { Router } from "express";
import { z } from "zod";
import { prisma } from "@guita-coach/db";
import { requireAuth, ah, HttpError, expenseAmount, reimbursementMap, daysInMonth } from "@guita-coach/shared";
import { applyRecurring } from "../services/recurring.js";
export const budgetRouter = Router();
budgetRouter.use(requireAuth);
const CATS = ["necesidades", "gustos", "ahorro"];
const LABELS = {
  necesidades: "Necesidades",
  gustos: "Gustos",
  ahorro: "Ahorro"
};
function franjaData(user, txs, month, daysRemaining = 1, reimb = {}, incomeBase) {
  const income = incomeBase ?? user.monthlyIncome ?? 0;
  const limits = {
    necesidades: income * user.necesidadesPct / 100,
    gustos: income * user.gustosPct / 100,
    ahorro: income * user.ahorroPct / 100
  };
  const visible = txs.filter(t => !t.isInternalTransfer && !t.isDuplicate);
  const spent = {};
  for (const cat of CATS) {
    spent[cat] = visible.filter(t => t.category === cat && (t.txType || "expense") === "expense").reduce((s, t) => s + expenseAmount(t, reimb), 0);
  }
  const dr = Math.max(daysRemaining, 1);
  const pctByCat = {
    necesidades: user.necesidadesPct,
    gustos: user.gustosPct,
    ahorro: user.ahorroPct
  };
  return {
    month,
    income,
    franjas: CATS.map(cat => ({
      name: cat,
      label: LABELS[cat],
      pct_config: pctByCat[cat],
      limit: limits[cat],
      spent: spent[cat],
      remaining: Math.max(0, limits[cat] - spent[cat]),
      usage_pct: limits[cat] > 0 ? Math.round(spent[cat] / limits[cat] * 1000) / 10 : 0,
      daily_allowance: Math.round(Math.max(0, limits[cat] - spent[cat]) / dr)
    }))
  };
}
budgetRouter.get("/current", ah(async (req, res) => {
  const user = req.user;
  const now = new Date();
  const currentMonth = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, "0")}`;
  let month = String(req.query.month || "");
  let target;
  if (month) {
    const m = month.match(/^(\d{4})-(\d{2})$/);
    if (!m) throw new HttpError(400, "Mes inválido, usá el formato YYYY-MM");
    target = new Date(Number(m[1]), Number(m[2]) - 1, 1);
    month = `${m[1]}-${m[2]}`;
  } else {
    month = currentMonth;
    target = new Date(now.getFullYear(), now.getMonth(), 1);
  }
  const isCurrent = month === currentMonth;
  if (isCurrent) await applyRecurring(user.id);
  const txs = await prisma.transaction.findMany({
    where: {
      userId: user.id,
      month,
      status: {
        in: ["confirmed", "classified"]
      }
    }
  });
  const days = daysInMonth(target.getFullYear(), target.getMonth() + 1);
  const daysPassed = isCurrent ? now.getDate() : days;
  const daysRemaining = isCurrent ? Math.max(days - daysPassed, 1) : 0;
  const reimb = await reimbursementMap(user.id);
  const visible = txs.filter(t => !t.isInternalTransfer && !t.isDuplicate);
  const trackedIncome = visible.filter(t => (t.txType || "expense") === "income" && !t.isReimbursement).reduce((s, t) => s + (t.amount ?? 0), 0);
  const totalExpenses = visible.filter(t => (t.txType || "expense") === "expense").reduce((s, t) => s + expenseAmount(t, reimb), 0);
  const isVariable = Boolean(user.incomeIsVariable);
  const declaredIncome = user.monthlyIncome ?? 0;
  let totalIncome;
  let incomeBase;
  if (isVariable) {
    totalIncome = trackedIncome;
    incomeBase = trackedIncome;
  } else {
    totalIncome = Math.max(trackedIncome, declaredIncome);
    incomeBase = totalIncome;
  }
  const data = franjaData(user, txs, month, daysRemaining, reimb, incomeBase);
  data.days_passed = daysPassed;
  data.days_in_month = days;
  data.days_remaining = days - daysPassed;
  const pendingCount = txs.filter(t => t.needsReview && t.status !== "reviewed").length;
  data.month = month;
  data.is_current_month = isCurrent;
  data.total_income = totalIncome;
  data.tracked_income = trackedIncome;
  data.declared_income = declaredIncome;
  data.income_is_variable = isVariable;
  data.income_is_declared = !isVariable && declaredIncome > trackedIncome;
  data.total_expenses = totalExpenses;
  data.balance = user.balance ?? 0;
  data.pending_count = pendingCount;
  data.onboarding_done = user.onboardingDone;
  data.name = user.name;
  data.payday = user.payday;
  const alerts = await prisma.alert.findMany({
    where: {
      userId: user.id,
      isRead: false
    },
    orderBy: {
      createdAt: "desc"
    }
  });
  data.alerts = alerts.map(a => ({
    id: a.id,
    type: a.type,
    category: a.category,
    message: a.message,
    ai_advice: a.aiAdvice,
    severity: a.severity,
    payload: a.payload,
    created_at: a.createdAt.toISOString()
  }));
  res.json(data);
}));
const OnboardingSchema = z.object({
  name: z.string(),
  monthly_income: z.number(),
  necesidades_pct: z.number().default(50),
  gustos_pct: z.number().default(30),
  ahorro_pct: z.number().default(20),
  payday: z.number().default(1),
  income_is_variable: z.boolean().default(false)
});
budgetRouter.post("/onboarding", ah(async (req, res) => {
  const payload = OnboardingSchema.parse(req.body);
  if (payload.necesidades_pct + payload.gustos_pct + payload.ahorro_pct !== 100) {
    throw new HttpError(400, "Los porcentajes deben sumar 100");
  }
  await prisma.user.update({
    where: {
      id: req.user.id
    },
    data: {
      name: payload.name,
      monthlyIncome: payload.monthly_income,
      necesidadesPct: payload.necesidades_pct,
      gustosPct: payload.gustos_pct,
      ahorroPct: payload.ahorro_pct,
      payday: payload.payday,
      incomeIsVariable: payload.income_is_variable,
      onboardingDone: true
    }
  });
  res.json({
    ok: true
  });
}));
const BudgetUpdateSchema = z.object({
  monthly_income: z.number().optional(),
  necesidades_pct: z.number().optional(),
  gustos_pct: z.number().optional(),
  ahorro_pct: z.number().optional(),
  payday: z.number().optional(),
  income_is_variable: z.boolean().optional()
});
budgetRouter.patch("/settings", ah(async (req, res) => {
  const payload = BudgetUpdateSchema.parse(req.body);
  const user = req.user;
  const data = {};
  if (payload.monthly_income !== undefined) data.monthlyIncome = payload.monthly_income;
  if (payload.necesidades_pct !== undefined) data.necesidadesPct = payload.necesidades_pct;
  if (payload.gustos_pct !== undefined) data.gustosPct = payload.gustos_pct;
  if (payload.ahorro_pct !== undefined) data.ahorroPct = payload.ahorro_pct;
  if (payload.payday !== undefined) data.payday = payload.payday;
  if (payload.income_is_variable !== undefined) data.incomeIsVariable = payload.income_is_variable;
  const necesidades = data.necesidadesPct ?? user.necesidadesPct;
  const gustos = data.gustosPct ?? user.gustosPct;
  const ahorro = data.ahorroPct ?? user.ahorroPct;
  const total = necesidades + gustos + ahorro;
  if (total !== 100) throw new HttpError(400, `Los porcentajes suman ${total}, deben ser 100`);
  await prisma.user.update({
    where: {
      id: user.id
    },
    data
  });
  res.json({
    ok: true
  });
}));
const BalanceSchema = z.object({
  balance: z.number()
});
budgetRouter.patch("/balance", ah(async (req, res) => {
  const payload = BalanceSchema.parse(req.body);
  const user = await prisma.user.update({
    where: {
      id: req.user.id
    },
    data: {
      balance: payload.balance
    }
  });
  res.json({
    ok: true,
    balance: user.balance
  });
}));
budgetRouter.get("/months", ah(async (req, res) => {
  const rows = await prisma.transaction.findMany({
    where: {
      userId: req.user.id,
      status: {
        in: ["confirmed", "classified"]
      }
    },
    select: {
      month: true
    },
    distinct: ["month"]
  });
  const months = new Set(rows.map(r => r.month).filter(Boolean));
  const now = new Date();
  months.add(`${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, "0")}`);
  res.json([...months].sort().reverse());
}));
budgetRouter.post("/alerts/:alertId/read", ah(async (req, res) => {
  const alert = await prisma.alert.findFirst({
    where: {
      id: Number(req.params.alertId),
      userId: req.user.id
    }
  });
  if (!alert) throw new HttpError(404, "Alerta no encontrada");
  await prisma.alert.update({
    where: {
      id: alert.id
    },
    data: {
      isRead: true
    }
  });
  res.json({
    ok: true
  });
}));
budgetRouter.get("/history", ah(async (req, res) => {
  const user = req.user;
  const rows = await prisma.transaction.findMany({
    where: {
      userId: user.id,
      status: {
        in: ["confirmed", "classified"]
      }
    },
    select: {
      month: true
    },
    distinct: ["month"]
  });
  const months = [...new Set(rows.map(r => r.month).filter(Boolean))].sort().reverse().slice(0, 6);
  const reimb = await reimbursementMap(user.id);
  const result = [];
  for (const month of months) {
    const txs = await prisma.transaction.findMany({
      where: {
        userId: user.id,
        month,
        status: {
          in: ["confirmed", "classified"]
        }
      }
    });
    result.push(franjaData(user, txs, month, 1, reimb));
  }
  res.json(result);
}));
